Tests/kaas/kaas-sonobuoy-tests/scs_k8s_conformance_tests/scs_0217_cluster_hardening_test.go

```diff
@@ -0,0 +1,825 @@
+				}
+			}
+			return ctx
+		})
+
+	testenv.Test(t, f.Feature())
+}
+
+// Test_Control_Plane_Ports_Security checks if the control plane ports (other than API server and NodePorts)
+// are not accessible from outside the internal network.
+func Test_scs_0217_sonobuoy_Control_Plane_Ports_Security(t *testing.T) {
+	f := features.New("control plane security").Assess(
+		"Control plane ports (other than API server and NodePorts) should not be accessible externally",
+		func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
+			nodes := &corev1.NodeList{}
+			labelSelector := labels.Set{
+				"node-role.kubernetes.io/control-plane": "",
+			}.AsSelector().String()
+
+			err := cfg.Client().Resources().List(context.TODO(), nodes, resources.WithLabelSelector(labelSelector))
+			if err != nil {
+				t.Fatal("failed to list control plane nodes:", err)
+			}
+
+			for _, node := range nodes.Items {
+				nodeIP := node.Status.Addresses[0].Address
+
+				// Check that the API server port (6443) is accessible
+				checkPortAccessibility(t, nodeIP, apiServerPort, true)
+
+				// Check that the control plane ports (other than API server) are not accessible
+				checkPortAccessibility(t, nodeIP, controllerManagerPort, false)
+				checkPortAccessibility(t, nodeIP, schedulerPort, false)
+				checkPortAccessibility(t, nodeIP, kubeletApiPort, false)
+
+				// Check the etcd ports (2379-2380) are not accessible
+				for port := etcdPortRangeStart; port <= etcdPortRangeEnd; port++ {
+					checkPortAccessibility(t, nodeIP, fmt.Sprintf("%d", port), false)
+				}
+			}
+			return ctx
+		})
+
+	testenv.Test(t, f.Feature())
+}
+
+// Test_K8s_Endpoints_HTTPS checks if all Kubernetes endpoints are secured via HTTPS.
+func Test_scs_0217_sonobuoy_K8s_Endpoints_HTTPS(t *testing.T) {
+	f := features.New("Kubernetes endpoint security").Assess(
+		"All Kubernetes endpoints should be secured via HTTPS",
+		func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
+			// Get the list of endpoints in the cluster
+			endpoints := &corev1.EndpointsList{}
+			err := cfg.Client().Resources().List(context.TODO(), endpoints)
+			if err != nil {
+				t.Fatal("failed to list endpoints:", err)
+			}
+
+			// Iterate over the endpoints and check if they're secured via HTTPS
+			for _, ep := range endpoints.Items {
+				for _, subset := range ep.Subsets {
+					for _, addr := range subset.Addresses {
+						nodeIP := addr.IP
+						for _, port := range subset.Ports {
+							portName := port.Name
+							portNum := port.Port
+
+							// Check if the endpoint is secured via HTTPS
+							if isPortSecuredWithHTTPS(nodeIP, portNum, connectionTimeout) {
+								t.Logf("Endpoint %s:%d (%s) is secured via HTTPS", nodeIP, portNum, portName)
+							} else {
+								t.Errorf("Error: Endpoint %s:%d (%s) is not secured via HTTPS", nodeIP, portNum, portName)
+							}
+						}
+					}
+				}
+			}
+			return ctx
+		})
+
+	testenv.Test(t, f.Feature())
+}
```

Comment:
In order to make this test work on my cluster,  needed to change a few things:
Added required HTTPS ports map:
```
goCopyrequiredHTTPSPorts := map[string]bool{
    apiServerPort:         true,
    controllerManagerPort: true, 
    schedulerPort:         true,
    kubeletApiPort:        true,
}
```
This ensures only the ports are checked that must use HTTPS per the standard.

Added etcd ports dynamically:
```
goCopyfor port := etcdPortRangeStart; port <= etcdPortRangeEnd; port++ {
    requiredHTTPSPorts[strconv.Itoa(port)] = true
}
```
Handles the etcd port range (2379-2380).

Modified endpoint check loop to only test required ports:
```
goCopyportStr := strconv.Itoa(int(port.Port))
if requiredHTTPSPorts[portStr] {
    // HTTPS check
}
```
This excludes ports like DNS (53) and metrics (9153) that don't require HTTPS.

Fixed context usage (not needed for passing the test, but TODO should now be there):
```
goCopycontext.Background() instead of context.TODO()
```
My suggestion (complete code):
```
// Test_K8s_Endpoints_HTTPS checks if all Kubernetes endpoints are secured via HTTPS.
func Test_scs_0217_sonobuoy_K8s_Endpoints_HTTPS(t *testing.T) {
   f := features.New("Kubernetes endpoint security").Assess(
       "All Kubernetes endpoints should be secured via HTTPS",
       func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
           requiredHTTPSPorts := map[string]bool{
               apiServerPort:         true,
               controllerManagerPort: true, 
               schedulerPort:         true,
               kubeletApiPort:        true,
           }

           // Add etcd ports
           for port := etcdPortRangeStart; port <= etcdPortRangeEnd; port++ {
               requiredHTTPSPorts[strconv.Itoa(port)] = true
           }

           endpoints := &corev1.EndpointsList{}
           if err := cfg.Client().Resources().List(context.Background(), endpoints); err != nil {
               t.Fatal("failed to list endpoints:", err)
           }

           for _, ep := range endpoints.Items {
               for _, subset := range ep.Subsets {
                   for _, addr := range subset.Addresses {
                       for _, port := range subset.Ports {
                           portStr := strconv.Itoa(int(port.Port))
                           if requiredHTTPSPorts[portStr] {
                               if isPortSecuredWithHTTPS(addr.IP, port.Port, connectionTimeout) {
                                   t.Logf("Endpoint %s:%d (%s) is secured via HTTPS", addr.IP, port.Port, port.Name)
                               } else {
                                   t.Errorf("Error: Endpoint %s:%d (%s) is not secured via HTTPS", addr.IP, port.Port, port.Name)
                               }
                           }
                       }
                   }
               }
           }
           return ctx
       })
   testenv.Test(t, f.Feature())
}
```